### PR TITLE
Finalize Pods if StatefulSet is not found.

### LINKED ISSUE #4138: [Flaky test] StatefulSet integration when StatefulSet created should allow to update the PodTemplate in StatefulSet
/kind flake 

**What happened**:

The test failed on unrelated branch: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/4131/pull-kueue-test-e2e-main-1-29/1886439313734897664

**What you expected to happen**:

No failures

**How to reproduce it (as minimally and precisely as possible)**:

Repeat

**Anything else we need to know?**:

```
{Timed out after 10.001s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/util/util.go:102 with:
Error matcher expects an error.  Got:
    <nil>: nil failed [FAILED] Timed out after 10.001s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/util/util.go:102 with:
Error matcher expects an error.  Got:
    <nil>: nil
In [AfterEach] at: /home/prow/go/src/sigs.k8s.io/kueue/test/e2e/singlecluster/statefulset_test.go:83 @ 02/03/25 15:47:42.977
}
```

Also another test " StatefulSet integration when StatefulSet created should delete all pods on scale down to zero" failed with 
```
{Expected success, but got an error:
    <*errors.StatusError | 0xc0002cd4a0>: 
    resourceflavors.kueue.x-k8s.io "sts-rf" already exists
    {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "resourceflavors.kueue.x-k8s.io \"sts-rf\" already exists",
            Reason: "AlreadyExists",
            Details: {
                Name: "sts-rf",
                Group: "kueue.x-k8s.io",
                Kind: "resourceflavors",
                UID: "",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 409,
        },
    } failed [FAILED] Expected success, but got an error:
    <*errors.StatusError | 0xc0002cd4a0>: 
    resourceflavors.kueue.x-k8s.io "sts-rf" already exists
    {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "resourceflavors.kueue.x-k8s.io \"sts-rf\" already exists",
            Reason: "AlreadyExists",
            Details: {
                Name: "sts-rf",
                Group: "kueue.x-k8s.io",
                Kind: "resourceflavors",
                UID: "",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 409,
        },
    }
In [BeforeEach] at: /home/prow/go/src/sigs.k8s.io/kueue/test/e2e/singlecluster/statefulset_test.go:64 @ 02/03/25 15:47:42.995
}
```
I suppose this is a consequence of the first failure.

### LINKED ISSUE #4160: [Flaky E2E] StatefulSet integration when StatefulSet created should allow to change queue name if ReadyReplicas=0
/kind flake

**What happened**:

End To End Suite: kindest/node:v1.32.0: [It] StatefulSet integration when StatefulSet created should allow to change queue name if ReadyReplicas=0 

```
{Timed out after 45.001s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/e2e/singlecluster/statefulset_test.go:366 with:
Expected
    <int32>: 0
to equal
    <int32>: 3 failed [FAILED] Timed out after 45.001s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/e2e/singlecluster/statefulset_test.go:366 with:
Expected
    <int32>: 0
to equal
    <int32>: 3
In [It] at: /home/prow/go/src/sigs.k8s.io/kueue/test/e2e/singlecluster/statefulset_test.go:367 @ 02/06/25 10:58:59.115
}
```

**What you expected to happen**:
No failures

**How to reproduce it (as minimally and precisely as possible)**:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/4159/pull-kueue-test-e2e-main-1-32/1887452310766882816

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Finalize Pods if StatefulSet is not found.

#### Which issue(s) this PR fixes:

Fixes #4160
Fixes #4138

#### Does this PR introduce a user-facing change?

```release-note
Fix a bug that doesn't allow Kueue to delete Pods after a StatefulSet is deleted.
```

### CODE REVIEW

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 1]
```diff
@@ -48,54 +57,59 @@ type Reconciler struct {
...
-	if err := r.client.List(ctx, podList, client.InNamespace(sts.Namespace), client.MatchingLabels{
-		podcontroller.GroupNameLabel: GetWorkloadName(sts.Name),
+	if err := r.client.List(ctx, podList, client.InNamespace(req.Namespace), client.MatchingLabels{
+		podcontroller.GroupNameLabel: GetWorkloadName(req.Name),
 	}); err != nil {
 		return err
 	}
+
+	sts := &appsv1.StatefulSet{}
+	err := r.client.Get(ctx, req.NamespacedName, sts)
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:77]: I think this does not work well if Kueue is restarted after the reconcile event was started and before the Reconciler happened. 

The similar problem is solved in [k8s Job controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go) which also needs to clean finalizers of orphan pods and is resilient to restarts. We use the "ADDED" event to Pods to delete the orhaned pods. PTAL.

The solution in k8s might be hard to implement, but I think in Kueue we could relatively easliy add an additional recnciler for Pods owned by statefulSet. When the StatefulSet is gone then finalize the Pods.

**Comment by mimowo** [comment]
/hold
To make the solution resilient in case of kueue restarts https://github.com/kubernetes-sigs/kueue/pull/4150#discussion_r1946424978

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 1 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:77]: > I think this does not work well if Kueue is restarted after the reconcile event was started and before the Reconciler happened.

You're right, I missed handling create events. Fixed.

Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:77]: I also tested locally. It is working perfectly fine.

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 2]
```diff
@@ -144,15 +159,53 @@ func (r *Reconciler) Update(e event.UpdateEvent) bool {
...
+func (h *podHandler) Generic(context.Context, event.GenericEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	// To watch for pod ADDED events.
+	h.handle(e.Object, q)
+}
+
+func (h *podHandler) Update(_ context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	h.handle(e.ObjectNew, q)
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:188]: Do we need this? What is the scenario? IIUC as long as STS exists the pod updates are reflected in the STS status updates, and so STS should reconcile.

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 3]
```diff
@@ -144,15 +159,53 @@ func (r *Reconciler) Update(e event.UpdateEvent) bool {
...
+}
+
+func (h *podHandler) handle(obj client.Object, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod || pod.Annotations[podcontroller.SuspendedByParentAnnotation] != FrameworkName {
+		return
+	}
+	if controllerRef := metav1.GetControllerOf(pod); controllerRef != nil {
+		if controllerRef.Kind != gvk.Kind || controllerRef.APIVersion != gvk.GroupVersion().String() {
+			// The pod is controlled by an owner that is not an apps/v1 StatefulSet.
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:201]: The owner of the Pod is never StatefulSet, so I'm confused by the comment actually.

I think [here](https://github.com/kubernetes-sigs/kueue/blob/577474373fafcd953faae122c7797a5ca2a7dad4/pkg/controller/jobs/pod/pod_webhook.go#L173-L186) is the analogous problem solved properly. It would be nice to commonize the code.

**Review by mimowo — COMMENTED** [review]
It is going in the right direction. 

Left some comments, but it would be great to get an additional pair of eyes. Maybe @mszadkow or @dgrove-oss could make a pass too.

File: test/e2e/singlecluster/statefulset_test.go [hunk 4]
```diff
@@ -372,5 +380,50 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
 				gomega.Expect(k8sClient.Get(ctx, wlLookupKey, createdWorkload)).To(gomega.Succeed())
 			})
 		})
+
+		ginkgo.It("should delete all Pods if StatefulSet was deleted after being partially ready", func() {
```
Comment (mimowo) [inline test/e2e/singlecluster/statefulset_test.go:384]: Actually, how is this test related to the issue? Is the new assert "Check all pods are deleted" not enough to repro the issue?

Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go:384]: To reliably catch this error, we can check if the StatefulSet was deleted after being partially ready. However, it can also happens when we just delete StatefulSet.

Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go:384]: > So the new check "Check all pods are deleted" does not catch the issue reliably?

Yes, exactly.

>  I'm not seeing a difference between partially ready StatefulSet and fully ready. Do you know why it matters for the change of catching the issue?

To be honest, I'm not sure why. But on this way I'm catching an error everytime.

File: test/e2e/singlecluster/statefulset_test.go [hunk 5]
```diff
@@ -372,5 +380,50 @@ var _ = ginkgo.Describe("StatefulSet integration", func() {
...
+
+			ginkgo.By("Check workload is admitted", func() {
+				util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, createdWorkload)
+			})
+
+			ginkgo.By("Waiting for replicas is partially ready", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					createdStatefulSet := &appsv1.StatefulSet{}
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(statefulSet), createdStatefulSet)).To(gomega.Succeed())
+					g.Expect(createdStatefulSet.Status.ReadyReplicas).To(gomega.Equal(int32(3)))
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply test/e2e/singlecluster/statefulset_test.go]: This is good point. Thanks!

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Yeah, I think, you're right. We don't need to handle Pod updates.

File: test/e2e/singlecluster/statefulset_test.go (hunk 4 above)
Comment (mimowo) [reply test/e2e/singlecluster/statefulset_test.go:384]: I see, I'm ok to keep it then

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 6]
```diff
@@ -144,15 +159,52 @@ func (r *Reconciler) Update(e event.UpdateEvent) bool {
...
+var _ handler.EventHandler = (*podHandler)(nil)
+
+type podHandler struct{}
+
+func (h *podHandler) Generic(context.Context, event.GenericEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	// To watch for pod ADDED events.
+	h.handle(e.Object, q)
```
Comment (tenzen-y) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:184]: Couldn't we directly implement logic here instead of separated `handle` function?

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 7]
```diff
@@ -144,15 +159,52 @@ func (r *Reconciler) Update(e event.UpdateEvent) bool {
...
+}
+
+func (h *podHandler) Update(_ context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
```
Comment (tenzen-y) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:187]:
```suggestion
func (h *podHandler) Update(context.Context, event.UpdateEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
```

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go
Comment (tenzen-y) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:1]: Should we add a test case to verify if Pods owned by non StatefulSet (ex. ReplicaSet) are not handled by new mechanism. I think that we can just checking if the Pods are not deleted.

Additionally, could you add a another case if Pods with `kueue.x-k8s.io/pod-suspending-parent: XXX` (XXX indicates non StatefulSet) owned StatefulSet are not handled by this reconcile?

**Review by tenzen-y — COMMENTED** [review]
Do we need to implement similar reconciler in other frameworks depended on the Pod integrations?

**Comment by mbobrovskyi** [comment]
> Do we need to implement similar reconciler in other frameworks depended on the Pod integrations?

Yeah, we also need to implement the same logic in LWS. I'll prepare a follow-up PR.

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 6 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Done

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:1]: Done

Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:1]: Done

**Comment by tenzen-y** [comment]
> > Do we need to implement similar reconciler in other frameworks depended on the Pod integrations?
> 
> Yeah, we also need to implement the same logic in LWS. I'll prepare a follow-up PR.

Awesome!

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 8]
```diff
@@ -18,25 +18,34 @@ package statefulset
...
 
 	"sigs.k8s.io/kueue/pkg/controller/jobframework"
 	podcontroller "sigs.k8s.io/kueue/pkg/controller/jobs/pod"
 	clientutil "sigs.k8s.io/kueue/pkg/util/client"
-	"sigs.k8s.io/kueue/pkg/util/parallelize"
 	utilpod "sigs.k8s.io/kueue/pkg/util/pod"
 )
 
+const (
+	nodeBatchPeriod = time.Second
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:46]:
```suggestion
	syncBatchPeriod = time.Second
```

**Comment by mimowo** [comment]
/hold
I believe this comment needs more work / investigation: https://github.com/kubernetes-sigs/kueue/pull/4150/files#r1946593745

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 3 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:193]: I created a StatefulSet without Kueue and observed the same behavior:

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2025-02-10T11:56:26Z"
  generateName: nginx-statefulset-
  labels:
    app: nginx
    apps.kubernetes.io/pod-index: "0"
    controller-revision-hash: nginx-statefulset-89458d665
    statefulset.kubernetes.io/pod-name: nginx-statefulset-0
  name: nginx-statefulset-0
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: StatefulSet
    name: nginx-statefulset
    uid: 37acd725-d7ea-4039-8ad0-1c78d375e65c
  resourceVersion: "513"
  uid: a0d1f759-be3b-4b2b-a96f-894018a67125
```

We are not setting ownerReference on the Pod.

Comment (mimowo) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:193]: I see, indeed, it seems ReplicaSets are only used for Deployements and not for StatefulSets, so all good.

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go [hunk 9]
```diff
@@ -48,54 +57,62 @@ type Reconciler struct {
...
+	}
+
 	return r.finalizePods(ctx, sts, podList.Items)
 }
 
 func (r *Reconciler) finalizePods(ctx context.Context, sts *appsv1.StatefulSet, pods []corev1.Pod) error {
+	errs := make([]error, 0, len(pods))
+	for _, pod := range pods {
+		errs = append(errs, r.finalizePod(ctx, sts, &pod))
+	}
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler.go:96]: I think we can run in parallel as in other places removing finalizers, like [here](https://github.com/kubernetes-sigs/kueue/blob/3755065cfde22ef480b882e7558aa8cd1bee58de/pkg/controller/jobs/pod/pod_controller.go#L529-L534)

**Comment by mimowo** [comment]
/hold cancel
> I believe this comment needs more work / investigation: https://github.com/kubernetes-sigs/kueue/pull/4150/files#r1946593745

Actually, the pod is indeed owned directly by STS (tested without Kueue)

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 8 above)
Comment (mbobrovskyi) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go]: Good point. Thanks!

**Comment by mszadkow** [comment]
I have nothing to add here, lgtm

**Review by mimowo — COMMENTED** [review]
/lgtm
/approve
Thanks!
/cherry-pick release-0.10 release-0.9

**Comment by mimowo** [comment]
@mbobrovskyi please follow up with an analogous fix for LWS.

**Comment by mimowo** [comment]
@mbobrovskyi please also prepare the cherry-picks manually

File: pkg/controller/jobs/statefulset/statefulset_reconciler.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/controller/jobs/statefulset/statefulset_reconciler.go:198]: Yes, the relationship for each resources are

- Deployment -> ReplicaSet -> Pod
- StatefulSet -> Pod
- DaemonSet -> Pod
- Job -> Pod

**Comment by tenzen-y** [comment]
> SuspendedByParentAnnotation

Oh, I see. In that case, I'm ok without cherry-picking to release-0.9
@mimowo wdyt?

**Comment by mimowo** [comment]
I'm not sure, let's consider options for 0.9:
1. without this fix the StatefulSet integration has a severe issue which can result in orphaned pods when deleting a StatefulSet
2. if we cherry-pick without looking at `SuspendedByParentAnnotation` it will have worse performance, because the STS reconciler would fire for all STS (not only managed by Kueue)
3. somehow check if a pod belongs to STS managed by Kueue (maybe by the presence of pod-group annotation)?

out of the options 1 < 2 < 3. (3.) requires a bit of code that is not in the main branch, but I think it is still better than (1.) and (2.).

**Comment by tenzen-y** [comment]
I think that key point which aspect is more valuable (performance vs leaving orphan pods issue).
TBH, I'm leaning toward (2.) since hundreds of orphan pods will affect the cluster performance like api-server, etcd...

**Comment by mimowo** [comment]
Right, to me correctness is also more important than performance. However, I believe with (3.) we can have both. 
Just check if the Kueue-specific pod-group annotation is present on the pod (and check if the StatefulSet if ownerReference). With these two constraints the performance will be good and behavior correct. 

I think the check for `SuspendedByParentAnnotation` on the main branch is only marginally better to skip reconciliation of pods owned by LWS via StatefulSet. But since 0.9 does not support LWS then looking for pod-group annotation is perfectly fine IMO.

**Comment by tenzen-y** [comment]
> However, I believe with (3.) we can have both. Just check if the Kueue-specific pod-group annotation is present on the pod (and check if the StatefulSet if ownerReference). With these two constraints the performance will be good and behavior correct.
> 
> I think the check for `SuspendedByParentAnnotation` on the main branch is only marginally better to skip reconciliation of pods owned by LWS via StatefulSet. But since 0.9 does not support LWS then looking for pod-group annotation is perfectly fine IMO.

Yeah, one more problem is conflicts. What if we select (3.)? I want to compare the amount of conflicts between release-0.9 branch and main.
If the about conflicts are mostly same, it would be great to select (3.), but if (3.) brings more conflicts, I'd like to select (2.).
@mimowo wdyt?

**Comment by mimowo** [comment]
I think the "fix" is very small, PTAL https://github.com/kubernetes-sigs/kueue/pull/4206#discussion_r1949363604

**Comment by tenzen-y** [comment]
> I think the "fix" is very small, PTAL [#4206 (comment)](https://github.com/kubernetes-sigs/kueue/pull/4206#discussion_r1949363604). In this comment I also proposed to add a comment which will help us to resolve conflicts in the future I believe by clearly indicating why the code is added.

commented in PR
